crates/astria-sequencer/src/app.rs

```diff
@@ -482,44 +679,24 @@ impl App {
     /// Note that the first two "transactions" in the block, which are the proposer-generated
     /// commitments, are ignored.
     #[instrument(name = "App::deliver_tx_after_proposal", skip_all, fields(
-        tx_hash =  %telemetry::display::base64(&Sha256::digest(&tx.tx)),
+        tx_hash =  %telemetry::display::base64(&Sha256::digest(tx)),
     ))]
     pub(crate) async fn deliver_tx_after_proposal(
         &mut self,
-        tx: abci::request::DeliverTx,
-    ) -> Option<anyhow::Result<Vec<abci::Event>>> {
-        self.current_sequencer_block_builder
-            .as_mut()
-            .expect(
-                "begin_block must be called before deliver_tx, thus \
-                 current_sequencer_block_builder must be set",
-            )
-            .push_transaction(tx.tx.to_vec());
-
-        if self.processed_txs < 2 {
-            self.processed_txs += 1;
-            return Some(Ok(vec![]));
-        }
-
+        tx: &bytes::Bytes,
+    ) -> anyhow::Result<Vec<abci::Event>> {
         // When the hash is not empty, we have already executed and cached the results
         if !self.executed_proposal_hash.is_empty() {
-            let tx_hash: [u8; 32] = sha2::Sha256::digest(&tx.tx).into();
-            return self.execution_result.remove(&tx_hash);
+            let tx_hash: [u8; 32] = sha2::Sha256::digest(tx).into();
+            return self
+                .execution_result
+                .remove(&tx_hash)
+                .unwrap_or_else(|| Err(anyhow!("transaction not executed during proposal phase")));
         }
 
-        let signed_tx = match signed_transaction_from_bytes(&tx.tx) {
-            Err(e) => {
-                // this is actually a protocol error, as only valid txs should be finalized
-                debug!(
-                    error = AsRef::<dyn std::error::Error>::as_ref(&e),
-                    "failed to decode deliver tx payload to signed transaction; ignoring it",
-                );
-                return None;
-            }
-            Ok(tx) => tx,
-        };
-
-        Some(self.deliver_tx(signed_tx).await)
+        let signed_tx = signed_transaction_from_bytes(tx)
```

Comment:
Couldn't this just be `signed_transaction_from_bytes(tx)?;`?  We're providing context to the error inside `signed_transaction_from_bytes`, and `deliver_tx_after_proposal` is already returning an `anyhow::Result`.

Reply:
you mean returning the error as opposed to panicking? this needs to be an `expect` as it's an unrecoverable protocol error, it would only happen if the chain's consensus failed to commit valid transactions. if we return an error here instead of panicking, the chain will halt anyways as cometbft will error due to the transaction results not matching, or some other error

Reply:
I think that's ok though?  If the higher logic knows to exit the process, or it causes the process to exit because cometbft dies then we have the correct outcome.

By panicking here we lose the source chain of the error - we'll only see the contents of the eyre `.context` value.  If we do want to panic, it might be better to do it inside `signed_transaction_from_bytes` (assuming there's never a case where that fails but we can continue to run) and we can replace `.context` with `.expect` to see the more relevant details.